[BUG] Replace numba with PyTorch for getting GPU info

### Description
<!--- Describe your issue/bug/request in detail -->
______________________________ test_get_gpu_info _______________________________

    def get_gpu_info():
        """Get information of GPUs.
    
        Returns:
            list: List of gpu information dictionary as with `device_name`, `total_memory` (in Mb) and `free_memory` (in Mb).
            Returns an empty list if there is no cuda device available.
        """
        gpus = []
        try:
            for gpu in cuda.gpus:
                with gpu:
>                   meminfo = cuda.current_context().get_memory_info()

recommenders/utils/gpu_utils.py:46: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../.venvs/Recommenders/lib/python3.9/site-packages/numba/cuda/cudadrv/devices.py:220: in get_context
    return _runtime.get_or_create_context(devnum)
../.venvs/Recommenders/lib/python3.9/site-packages/numba/cuda/cudadrv/devices.py:138: in get_or_create_context
    return self._get_or_create_context_uncached(devnum)
../.venvs/Recommenders/lib/python3.9/site-packages/numba/cuda/cudadrv/devices.py:153: in _get_or_create_context_uncached
    with driver.get_active_context() as ac:
../.venvs/Recommenders/lib/python3.9/site-packages/numba/cuda/cudadrv/driver.py:505: in __enter__
    driver.cuCtxGetDevice(byref(hdevice))
../.venvs/Recommenders/lib/python3.9/site-packages/numba/cuda/cudadrv/driver.py:327: in safe_cuda_api_call
    self._check_ctypes_error(fname, retcode)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <numba.cuda.cudadrv.driver.Driver object at 0x7dfe2c8f8730>
fname = 'cuCtxGetDevice', retcode = 201

    def _check_ctypes_error(self, fname, retcode):
        if retcode != enums.CUDA_SUCCESS:
            errname = ERROR_MAP.get(retcode, "UNKNOWN_CUDA_ERROR")
            msg = "Call to %s results in %s" % (fname, errname)
            _logger.error(msg)
            if retcode == enums.CUDA_ERROR_NOT_INITIALIZED:
                self._detect_fork()
>           raise CudaAPIError(retcode, msg)
E           numba.cuda.cudadrv.driver.CudaAPIError: [201] Call to cuCtxGetDevice results in CUDA_ERROR_INVALID_CONTEXT


### In which platform does it happen?
<!--- Describe the platform where the issue is happening (use a list if needed) -->
<!--- For example: -->
<!--- * Azure Data Science Virtual Machine. -->
<!--- * Azure Databricks.  -->
<!--- * Other platforms.  -->

Python 3.9

### How do we replicate the issue?
<!--- Please be specific as possible (use a list if needed). -->
<!--- For example: -->
<!--- * Create a conda environment for pyspark -->
<!--- * Run unit test `test_sar_pyspark.py` with `pytest -m 'spark'` -->
<!--- * ... -->

### Expected behavior (i.e. solution)
<!--- For example:  -->
<!--- * The tests for SAR PySpark should pass successfully. -->

### Willingness to contribute
<!--- Go over all the following points, and put an `x` in the box that apply. -->
- [ ] Yes, I can contribute for this issue independently.
- [ ] Yes, I can contribute for this issue with guidance from Recommenders community.
- [ ] No, I cannot contribute at this time.

### Other Comments

--- a/recommenders/utils/gpu_utils.py
+++ b/recommenders/utils/gpu_utils.py
@@ -8,7 +8,6 @@ import logging
 from numba import cuda
 from numba.cuda.cudadrv.error import CudaSupportError
 
-
 logger = logging.getLogger(__name__)
 
 
@@ -41,6 +40,23 @@ def get_gpu_info():
     """
     gpus = []
     try:
+        import torch
+
+        if torch.cuda.is_available():
+            for device_id in range(torch.cuda.device_count()):
+                free_memory, total_memory = torch.cuda.mem_get_info(device_id)
+                gpus.append(
+                    {
+                        "device_name": torch.cuda.get_device_name(device_id),
+                        "total_memory": total_memory / 1048576,  # Mb
+                        "free_memory": free_memory / 1048576,  # Mb
+                    }
+                )
+            return gpus
+    except (ImportError, ModuleNotFoundError):
+        pass
+
+    try:
         for gpu in cuda.gpus:
             with gpu:
                 meminfo = cuda.current_context().get_memory_info()
@@ -50,7 +66,7 @@ def get_gpu_info():
                     "free_memory": meminfo[0] / 1048576,  # Mb
                 }
                 gpus.append(g)
-    except CudaSupportError:
+    except Exception:
         pass
 
     return gpus

--- a/tests/unit/recommenders/utils/test_gpu_utils.py
+++ b/tests/unit/recommenders/utils/test_gpu_utils.py
@@ -38,6 +38,50 @@ def test_get_number_gpus_without_torch(monkeypatch):
     assert gpu_utils.get_number_gpus() == 2
 
 
+def test_get_gpu_info_uses_torch_cuda(monkeypatch):
+    class FakeCuda:
+        @staticmethod
+        def is_available():
+            return True
+
+        @staticmethod
+        def device_count():
+            return 1
+
+        @staticmethod
+        def mem_get_info(device):
+            assert device == 0
+            return 256 * 1048576, 1024 * 1048576
+
+        @staticmethod
+        def get_device_name(device):
+            assert device == 0
+            return "Test GPU"
+
+    class BrokenNumbaGpu:
+        def __enter__(self):
+            raise AssertionError("numba should not be used when torch has CUDA")
+
+    fake_torch = SimpleNamespace(cuda=FakeCuda)
+    real_import = builtins.__import__
+
+    def fake_import(name, globals=None, locals=None, fromlist=(), level=0):
+        if name == "torch":
+            return fake_torch
+        return real_import(name, globals, locals, fromlist, level)
+
+    monkeypatch.setattr(gpu_utils, "cuda", SimpleNamespace(gpus=[BrokenNumbaGpu()]))
+    monkeypatch.setattr(builtins, "__import__", fake_import)
+
+    assert gpu_utils.get_gpu_info() == [
+        {
+            "device_name": "Test GPU",
+            "total_memory": 1024,
+            "free_memory": 256,
+        }
+    ]
+
+
 @pytest.mark.gpu
 def test_get_gpu_info():
     assert len(get_gpu_info()) >= 1

